lib/mongo/server/connection.rb

```diff
@@ -173,11 +173,21 @@ def connect!
 
       # Separate method to permit easier mocking in the test suite.
       def do_connect
-        socket = address.socket(socket_timeout, ssl_options,
-          connect_timeout: address.connect_timeout)
-        handshake!(socket)
-        pending_connection = PendingConnection.new(socket, @server, monitoring, options)
-        authenticate!(pending_connection)
+        socket = nil
+
+        begin
+          socket = address.socket(socket_timeout, ssl_options,
+            connect_timeout: address.connect_timeout)
+          handshake!(socket)
+          pending_connection = PendingConnection.new(socket, @server, monitoring, options)
+          authenticate!(pending_connection)
+        rescue Exception => e
+          if socket
+            socket.close
+          end
+          raise e
```

Comment:
Ruby allows an argument-less `raise` which will reraise the exception being handled, which also would make it unnecessary to assign it to a variable (`=> e` part).